zkvm/src/vm.rs

```diff
@@ -306,6 +306,51 @@ where
         Ok(())
     }
 
+    fn mul(&mut self) -> Result<(), VMError> {
+        let expr2 = self.pop_item()?.to_expression()?;
+        let expr1 = self.pop_item()?.to_expression()?;
+
+        // Note: we cannot implement `mul` for `Expression` because we need to have access
+        // to the `ConstraintSystem` API in order to make a multiplier in the
+        // `LinearCombination * LinearCombination` case.
+        let expr3 = match (expr1, expr2) {
```

Comment:
You can factor this into `Expression::multiply<CS: ConstraintSystem>(self, rhs, cs: &mut CS)` since you don't need anything else but `CS::multiply`, as far as i can see.

Reply:
done! thanks, that's much better / more readable now